Sources/Apodini/Semantic Model Builder/EndpointsTree.swift

```diff
@@ -0,0 +1,172 @@
+    lazy var responseType: ResponseEncodable.Type = {
+        guard let lastResponseTransformer = self.responseTransformers.last else {
+            return self.handleReturnType
+        }
+        return lastResponseTransformer().transformedResponseType
+    }()
+    
+    /// All `@Parameter` `RequestInjectable`s that are used inside handling `Component`
+    lazy var parameters: [EndpointParameter] = EndpointParameter.create(from: requestInjectables)
+
+    var absolutePath: [_PathComponent] {
+        treeNode.absolutePath
+    }
+}
+
+class EndpointsTreeNode {
+    let path: _PathComponent
+    var endpoints: [Operation: Endpoint] = [:]
+
+    let parent: EndpointsTreeNode?
+    private var nodeChildren: [String: EndpointsTreeNode] = [:]
+    var children: Dictionary<String, EndpointsTreeNode>.Values {
+        nodeChildren.values
+    }
+    /// If a EndpointsTreeNode A is a child to  a EndpointsTreeNode B and A has an `PathParameter` as its `path`
+    /// B can't have any other children besides A that also have an `PathParameter` at the same location.
+    /// Thus we mark `childContainsPathParameter` to true as soon as we insert a `PathParameter` as a child.
+    private var childContainsPathParameter = false
+
+    lazy var absolutePath: [_PathComponent] = {
+        var absolutePath: [_PathComponent] = []
+        collectAbsolutePath(&absolutePath)
+        return absolutePath
+    }()
+
+    init(path: _PathComponent, parent: EndpointsTreeNode? = nil) {
+        self.path = path
+        self.parent = parent
+    }
+
+    func addEndpoint(_ endpoint: inout Endpoint, at paths: [PathComponent]) {
+        if paths.isEmpty {
+            // swiftlint:disable:next force_unwrapping
+            precondition(endpoints[endpoint.operation] == nil, "Tried overwriting endpoint \(endpoints[endpoint.operation]!.description) with \(endpoint.description) for operation \(endpoint.operation)")
+            precondition(endpoint.treeNode == nil, "The endpoint \(endpoint.description) is already inserted at some different place")
+
+            endpoint.treeNode = self
+            endpoints[endpoint.operation] = endpoint
+        } else {
+            var pathComponents = paths
+            if let first = pathComponents.removeFirst() as? _PathComponent {
+                var child = nodeChildren[first.description]
+                if child == nil {
+                    // swiftlint:disable:next todo
+                    // TODO is there a better way to check for type Parameter instead of string comparison?
+                    if let info = try? typeInfo(of: type(of: first)), info.mangledName == "Parameter" {
+                        let mirror = Mirror(reflecting: first)
+
+                        // swiftlint:disable:next force_cast
+                        let options = mirror.children.first { $0.label == "options" }!.value as! PropertyOptionSet<ParameterOptionNameSpace>
```

Comment:
I think we should add some enum with static properties that stores these constants such as `"options"` and `"Parameter"` where they are defined/in the `Parameter` type and use it here. Then we would at least have one file that stores all this context.

Reply:
Could you maybe illustrate how you want this to look? Just an empty enum with `static let` definitions?

Reply:
I hope 967e95c is what you had in mind?